docs/how-it-works.md

```diff
@@ -0,0 +1,55 @@
+# How it works
+
+The objective of this document explains how the NGINX Ingress controller works, in particular how the NGINX model is built and why we need a one.
+
+## NGINX configuration
+
+The goal of this Ingress controller is the assembly of a configuration file (nginx.conf). The main implication of this requirement is the need to reload NGINX after a change in the configuration file.
+
+## NGINX model
+
+Usually, a Kubernetes Controller utilizes the [synchronization loop pattern](1) to check if the desired state in the controller is updated or a change is required. To this purpose, we need to build a model using different objects from the cluster. This Ingress controller in particular watches (in no special order) Ingresses, Services, Endpoints, Secrets, and Configmaps to generate a point in time configuration file that reflects the state of the cluster.
+To get this object from the cluster we use [Kubernetes Informers](2), in particular `FilteredSharedInformer`. This informers allow reacting [callbacks](3) to individual changes when a new object is added, modified or removed. Unfortunately, there is no way to know if a particular change is going to affect the final configuration file. This is one of the uses of the model is avoid unnecessary reloads when there's no change in the state and to detect conflicts in definitions.
```

Comment:
Please check my other comment at https://github.com/kubernetes/ingress-nginx/pull/2479#discussion_r190893250. It was lost after your commit.